app/app.go

```diff
@@ -35,19 +36,40 @@ type RawObject struct {
 	Encoding resource.KindEncoding `json:"-"`
 }
 
-// ResourceCustomRouteRequest is a request to a custom subresource
-type ResourceCustomRouteRequest struct {
+// CustomRouteRequest is a request to a custom subresource or resource route
+type CustomRouteRequest struct {
+	// ResourceIdentifier is the full identifier of the resource.
+	// If the request is not a subresource route request, ResourceIdentifier will contain only group,
+	// version, and namespace (if applicable).
+	// ResourceIdentifier will contain all information the runner is able to provide.
+	// In practical terms, this often means that either Kind or Plural will be included, but both may not be present.
 	ResourceIdentifier resource.FullIdentifier
-	SubresourcePath    string
-	Method             string
-	Headers            http.Header
-	Body               []byte
+	// Path contains path information past the identifier information.
+	// For a subresource route, this is the subresource (for example, `bar` in the case of `test.grafana.app/v1/foos/foo/bar`).
+	// In the case of a non-subresource route, this will be the path section past the namespace (or version if the route is not namespaced).
+	Path string
+	// Method is the HTTP request method
+	Method string
+	// Headers contains the HTTP headers of the original request. Runners MAY remove or sanitize some headers.
+	Headers http.Header
+	// Body contains the payload of the request
+	Body []byte
 }
 
-type ResourceCustomRouteResponse struct {
-	Headers    http.Header
+// CustomRouteResponse is the response object to a CustomRouteRequest
+type CustomRouteResponse struct {
+	// Headers contains the HTTP headers for the response
+	Headers http.Header
+	// StatusCode if the HTTP status code of the response
 	StatusCode int
-	Body       []byte
+	// Body is the payload of the response. Runners MAY support response streaming.
+	// In this case, an EOF returned by Body.Read() should designate the end of the response
+	// (0 bytes read with no error only denotes that there is nothing currently to write to the response stream),
+	// consistent with the behavior described in the godoc for io.Reader.
+	// If a Runner does not support response streaming, it should call Close() on Body when the payload's response
+	// has been written and the stream closed (this should also be done when streaming is supported after getting an
+	// EOF or other error and no further bytes).
+	Body io.ReadCloser
```

Comment:
The other solution here (as pointed out by @toddtreece) would be to provide a writer/sender object in the request object or `CallCustomRoute` method, and the app can keep the response stream open by not returning from the method. I think that would move us more toward following the `http.ResponseWriter` approach, which we had initially and changed to a concrete response object. I'd love opinions on which of these two approaches we think may be better.

Reply:
the writer/sender approach seems slightly easier for users that only need to send `[]byte`, but wrapping with `io.NopCloser(bytes.NewReader([]byte{}))` isn't that huge of an issue, so either way works for me.

this seems like an edge case, but the only thing i can think of is that sending trailer headers wouldn't be supported with `io.ReadCloser` (although maybe you could write them to the body? not sure)